### ISSUE TITLE
Get value and gradient simultaneously

### ISSUE THREAD

**Issue by sscardapane** [google/jax#149]
In Autograd I was used to have `value_and_grad` to return both the gradient and the value of a function simultaneously. Is there any similar functionality in JAX? That would be incredibly useful for, e.g., plotting the loss while training a neural network.

This is an example in Autograd:

```
grad_and_value = grad_and_value(np.tanh)
v, g = grad_and_value(0.0) # v=0.0, g=1.0
```

**Comment by mattjj** [comment]
Great idea. If you take a look at [the grad function's implementation](https://github.com/google/jax/blob/95135377d0ed3d3954a44a650ad2580ded61037f/jax/api.py#L105), you can see that `ans` is computed but not returned. We should factor the API so that this is easy to access.

@dougalm do you think it's better to set things up so that there's a separate `value_and_grad` function (and maybe `grad` is a wrapped version that discards the `ans` part of its return value), or is it worth considering just having a `return_ans` kwarg for `grad`? My usual preference is for the former, since other APIs always seem to succumb to kwarg creep, but in this case I'm not sure.

**Comment by alexbw** [comment]
I would vote for a separate function. Having kwargs affect what is returned
to me is a Matlab anti-pattern, but if folx disagree, let's discuss.

**Comment by mattjj** [comment]
@dougalm agreed, and had a punchy way of saying the same thing @alexbw did: an argument to a function shouldn't change its type! That sounds like a knock-down argument to me.

I'll add a value_and_grad function now.

**Comment by mattjj** [comment]
Please reopen if there's something we missed here!
